File: apps/android/app/src/main/java/inc/anky/android/core/identity/RecoveryPhrase.kt
package inc.anky.android.core.identity

import java.security.MessageDigest
import java.security.SecureRandom
import org.web3j.crypto.MnemonicUtils

data class RecoveryPhrase(val words: List<String>) {
    val text: String = words.joinToString(separator = " ")

    init {
        require(words.size == 12) { "Recovery phrase must contain 12 words." }
        require(words.all { BIP39WordList.english.contains(it) }) { "Recovery phrase contains an unsupported word." }
    }

    fun signingSeed(): ByteArray = MnemonicUtils.generateSeed(text, "")

    /**
     * BIP39: the final 4 bits of a 12-word phrase are the top 4 bits of
     * SHA256 over its 128 entropy bits. Mirrors iOS RecoveryPhrase.hasValidChecksum
     * (ios/Anky/Core/Identity/RecoveryPhrase.swift).
     */
    val hasValidChecksum: Boolean
        get() {
            if (words.size != 12) return false
            val bits = BooleanArray(132)
            words.forEachIndexed { wordIndex, word ->
                val index = BIP39WordList.english.indexOf(word)
                if (index < 0) return false
                for (bit in 0 until 11) {
                    bits[wordIndex * 11 + bit] = (index shr (10 - bit)) and 1 == 1
                }
            }
            val entropy = ByteArray(16)
            for (position in 0 until 128) {
                if (bits[position]) {
                    entropy[position / 8] = (entropy[position / 8].toInt() or (1 shl (7 - position % 8))).toByte()
                }
            }
            val checksumByte = MessageDigest.getInstance("SHA-256").digest(entropy)[0].toInt() and 0xFF
            for (offset in 0 until 4) {
                val expected = (checksumByte shr (7 - offset)) and 1 == 1
                if (bits[128 + offset] != expected) return false
            }
            return true
        }

    companion object {
        fun parse(text: String): RecoveryPhrase =
            RecoveryPhrase(text.lowercase().trim().split(Regex("\\s+")).filter { it.isNotEmpty() })

        /**
         * For phrases typed by a human (import): a dictionary-valid one-word
         * typo must fail here instead of silently deriving a stranger wallet.
         * Stored phrases keep loading through the non-validating [parse] so a
         * pre-validation import can never brick an existing identity. Mirrors
         * iOS RecoveryPhrase(text:validatingChecksum:).
         */
        fun parse(text: String, validatingChecksum: Boolean): RecoveryPhrase {
            val phrase = parse(text)
            require(!validatingChecksum || phrase.hasValidChecksum) { "Recovery phrase checksum is invalid." }
            return phrase
        }

        fun generate(random: SecureRandom = SecureRandom()): RecoveryPhrase {
            val entropy = ByteArray(16)
            random.nextBytes(entropy)
            return parse(MnemonicUtils.generateMnemonic(entropy))
        }
    }
}

File: apps/ios/Anky/Core/Identity/RecoveryPhrase.swift
import CryptoKit
import Foundation
import Security

struct RecoveryPhrase: Hashable {
    let words: [String]

    var text: String {
        words.joined(separator: " ")
    }

    init(words: [String]) throws {
        guard words.count == 12 else {
            throw RecoveryPhraseError.invalidWordCount
        }
        let allowed = Set(BIP39WordList.english)
        guard words.allSatisfy({ allowed.contains($0) }) else {
            throw RecoveryPhraseError.unknownWord
        }
        self.words = words
    }

    init(text: String) throws {
        let words = text
            .lowercased()
            .split(whereSeparator: \.isWhitespace)
            .map(String.init)
        try self.init(words: words)
    }

    /// For phrases typed by a human (import): a dictionary-valid one-word
    /// typo must fail here instead of silently deriving a stranger wallet.
    /// Stored phrases keep loading through the non-validating inits so a
    /// pre-validation import can never brick an existing identity.
    init(text: String, validatingChecksum: Bool) throws {
        try self.init(text: text)
        if validatingChecksum, !hasValidChecksum {
            throw RecoveryPhraseError.invalidChecksum
        }
    }

    /// BIP39: the final 4 bits of a 12-word phrase are the top 4 bits of
    /// SHA256 over its 128 entropy bits.
    var hasValidChecksum: Bool {
        guard words.count == 12 else {
            return false
        }
        var bits = [Bool]()
        bits.reserveCapacity(132)
        for word in words {
            guard let index = BIP39WordList.english.firstIndex(of: word) else {
                return false
            }
            for bit in 0..<11 {
                bits.append((index >> (10 - bit)) & 1 == 1)
            }
        }
        var entropy = Data(count: 16)
        for (position, bit) in bits.prefix(128).enumerated() where bit {
            entropy[position / 8] |= 1 << (7 - (position % 8))
        }
        let checksumByte = Data(SHA256.hash(data: entropy))[0]
        for (offset, bit) in bits.suffix(4).enumerated() {
            let expected = (checksumByte >> (7 - offset)) & 1 == 1
            if bit != expected {
                return false
            }
        }
        return true
    }

    static func generate() throws -> RecoveryPhrase {
        var entropy = Data(count: 16)
        let status = entropy.withUnsafeMutableBytes { buffer in
            SecRandomCopyBytes(kSecRandomDefault, buffer.count, buffer.baseAddress!)
        }
        guard status == errSecSuccess else {
            throw RecoveryPhraseError.randomFailed
        }

        var bits = entropy.reduce(into: [Bool]()) { result, byte in
            result.append(contentsOf: (0..<8).map { bit in ((byte >> (7 - bit)) & 1) == 1 })
        }
        let checksum = Data(SHA256.hash(data: entropy))
        let checksumByte = checksum[checksum.startIndex]
        bits.append((checksumByte & 0b1000_0000) != 0)
        bits.append((checksumByte & 0b0100_0000) != 0)
        bits.append((checksumByte & 0b0010_0000) != 0)
        bits.append((checksumByte & 0b0001_0000) != 0)

        let words = stride(from: 0, to: bits.count, by: 11).map { offset in
            var index = 0
            for bit in bits[offset..<(offset + 11)] {
                index = (index << 1) | (bit ? 1 : 0)
            }
            return BIP39WordList.english[index]
        }
        return try RecoveryPhrase(words: words)
    }
}

enum RecoveryPhraseError: Error, Equatable {
    case invalidWordCount
    case unknownWord
    case invalidChecksum
    case randomFailed
}
